drivers/clock_control/clock_control_nrf2_fll16m.c

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nordic_nrf_fll16m
+
+#include "clock_control_nrf2_common.h"
+LOG_MODULE_DECLARE(clock_control_nrf2, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
+	     "multiple instances not supported");
+
+#define FLAG_HFXO_STARTED BIT(FLAGS_COMMON_BITS)
+
+/* TODO: get from devicetree. */
+static const uint16_t option_accuracy[] = { 20000, 5000, 100 };
+static const uint16_t max_accuracy = 100;
+
+struct fll16m_dev_data {
+	STRUCT_CLOCK_CONFIG(fll16m, ARRAY_SIZE(option_accuracy)) clk_cfg;
+	struct onoff_client hfxo_cli;
+};
+
+struct fll16m_dev_config {
+	uint32_t clock_frequency;
+};
+
+static void hfxo_cb(struct onoff_manager *mgr,
+		    struct onoff_client *cli,
+		    uint32_t state,
+		    int res)
+{
+	ARG_UNUSED(mgr);
+	ARG_UNUSED(state);
+
+	struct fll16m_dev_data *dev_data =
+		CONTAINER_OF(cli, struct fll16m_dev_data, hfxo_cli);
+
+	if (res < 0) {
+		(void)atomic_or(&dev_data->clk_cfg.flags, FLAG_HFXO_STARTED);
+		NRF_LRCCONF010->CLKCTRL[0].SRC = 2;
+	}
+
+	clock_config_update_end(&dev_data->clk_cfg, res);
+}
+
+static void fll16m_work_handler(struct k_work *work)
+{
+	const struct device *hfxo = DEVICE_DT_GET(DT_NODELABEL(hfxo));
+	struct fll16m_dev_data *dev_data =
+		CONTAINER_OF(work, struct fll16m_dev_data, clk_cfg.work);
+	uint32_t active_options;
+
+	active_options = clock_config_update_begin(work);
+	if (active_options & BIT(2)) {
```

Comment:
These magic bits in `active_options` are hard to follow in the review

Reply:
Yes, this will be reworked.